chain/transaction.go

```diff
@@ -359,6 +362,92 @@ func (t *Transaction) Marshal(p *codec.Packer) error {
 	return t.marshal(p)
 }
 
+func (t *Transaction) MarshalJSON() ([]byte, error) {
+	type txJSON struct {
+		ID      ids.ID `json:"id"`
+		Actions []byte `json:"actions"`
+		Auth    []byte `json:"auth"`
+	}
+
+	actionsPacker := codec.NewWriter(0, consts.NetworkSizeLimit)
+	err := t.Actions.marshalInto(actionsPacker)
+	if err != nil {
+		return nil, err
+	}
+	if err := actionsPacker.Err(); err != nil {
+		return nil, err
+	}
+	authPacker := codec.NewWriter(0, consts.NetworkSizeLimit)
+	authPacker.PackByte(t.Auth.GetTypeID())
+	t.Auth.Marshal(authPacker)
+	if err := authPacker.Err(); err != nil {
+		return nil, err
+	}
+
+	hexActions, err := codec.Bytes(actionsPacker.Bytes()).MarshalText()
+	if err != nil {
+		return nil, err
+	}
+	hexAuth, err := codec.Bytes(authPacker.Bytes()).MarshalText()
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(txJSON{
+		ID:      t.ID(),
+		Actions: hexActions,
+		Auth:    hexAuth,
+	})
+}
+
+func (t *Transaction) UnmarshalJSON(data []byte, parser Parser) error {
+	type txJSON struct {
+		ID      ids.ID `json:"id"`
+		Actions []byte `json:"actions"`
+		Auth    []byte `json:"auth"`
+	}
+
+	var tx txJSON
+	err := json.Unmarshal(data, &tx)
+	if err != nil {
+		return err
+	}
+
+	var actionsBytes, authBytes codec.Bytes
+	err = actionsBytes.UnmarshalText(tx.Actions)
+	if err != nil {
+		return fmt.Errorf("%w: cannot unmarshal actions text", err)
+	}
+	err = authBytes.UnmarshalText(tx.Auth)
+	if err != nil {
+		return fmt.Errorf("%w: cannot unmarshal auth text", err)
+	}
+
+	actionsReader := codec.NewReader(actionsBytes, consts.NetworkSizeLimit)
+	actions, err := UnmarshalActions(actionsReader, parser.ActionCodec())
+	if err != nil {
+		return err
+	}
+	if err := actionsReader.Err(); err != nil {
+		return fmt.Errorf("%w: actions packer", err)
+	}
+	authReader := codec.NewReader(authBytes, consts.NetworkSizeLimit)
+	auth, err := parser.AuthCodec().Unmarshal(authReader)
+	if err != nil {
+		return fmt.Errorf("%w: cannot unmarshal auth", err)
+	}
+	if err := authReader.Err(); err != nil {
+		return fmt.Errorf("%w: auth packer", err)
+	}
+
+	t.id = tx.ID
+	t.TransactionData = TransactionData{
+		Actions: actions,
+	}
+	t.Auth = auth
+	return nil
```

Comment:
Rather than trusting the received txID (which could be a mismatch), we should re-marshal the received tx based on the fields (Base, Actions, and Auth) and unmarshal it to populate the txID.

Ideally we can restrict the number of places where we construct a transaction, so that it's clear all of the required fields are always populated.

Right now the only place we create `Transaction` is in `Sign`. We could create a new function `NewTx(base, actions, auth)` that we call within `Sign` to keep it to a single place where we construct a signed transaction.